src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/Utils/TypeNameHelpers.cs

```diff
@@ -0,0 +1,187 @@
+                PrimitiveType.UInt16 => "System.UInt16",
+                PrimitiveType.Int32 => "System.Int32",
+                PrimitiveType.UInt32 => "System.UInt32",
+                PrimitiveType.Int64 => "System.Int64",
+                PrimitiveType.UInt64 => "System.UInt64",
+                PrimitiveType.Single => "System.Single",
+                PrimitiveType.Double => "System.Double",
+                PrimitiveType.Decimal => "System.Decimal",
+                PrimitiveType.TimeSpan => "System.TimeSpan",
+                PrimitiveType.DateTime => "System.DateTime",
+                PrimitiveType.String => "System.String",
+                ObjectPrimitiveType => "System.Object",
+                IntPtrPrimitiveType => "System.IntPtr",
+                UIntPtrPrimitiveType => "System.UIntPtr",
+                _ => throw new ArgumentOutOfRangeException(paramName: nameof(primitiveType), actualValue: primitiveType, message: null)
+            };
+
+            s_PrimitiveTypeNames[(int)primitiveType] = typeName = TypeName.Parse(fullName.AsSpan()).WithCoreLibAssemblyName();
+        }
+        return typeName;
+    }
+
+    internal static TypeName GetPrimitiveSZArrayTypeName(PrimitiveType primitiveType)
+    {
+        TypeName? typeName = s_PrimitiveSZArrayTypeNames[(int)primitiveType];
+        if (typeName is null)
+        {
+            s_PrimitiveSZArrayTypeNames[(int)primitiveType] = typeName = GetPrimitiveTypeName(primitiveType).MakeSZArrayTypeName();
+        }
+        return typeName;
+    }
+
+    internal static PrimitiveType GetPrimitiveType<T>()
+    {
+        if (typeof(T) == typeof(bool)) return PrimitiveType.Boolean;
+        else if (typeof(T) == typeof(byte)) return PrimitiveType.Byte;
+        else if (typeof(T) == typeof(sbyte)) return PrimitiveType.SByte;
+        else if (typeof(T) == typeof(char)) return PrimitiveType.Char;
+        else if (typeof(T) == typeof(short)) return PrimitiveType.Int16;
+        else if (typeof(T) == typeof(ushort)) return PrimitiveType.UInt16;
+        else if (typeof(T) == typeof(int)) return PrimitiveType.Int32;
+        else if (typeof(T) == typeof(uint)) return PrimitiveType.UInt32;
+        else if (typeof(T) == typeof(long)) return PrimitiveType.Int64;
+        else if (typeof(T) == typeof(ulong)) return PrimitiveType.UInt64;
+        else if (typeof(T) == typeof(float)) return PrimitiveType.Single;
+        else if (typeof(T) == typeof(double)) return PrimitiveType.Double;
+        else if (typeof(T) == typeof(decimal)) return PrimitiveType.Decimal;
+        else if (typeof(T) == typeof(DateTime)) return PrimitiveType.DateTime;
+        else if (typeof(T) == typeof(TimeSpan)) return PrimitiveType.TimeSpan;
+        else if (typeof(T) == typeof(string)) return PrimitiveType.String;
+        else if (typeof(T) == typeof(IntPtr)) return IntPtrPrimitiveType;
+        else if (typeof(T) == typeof(UIntPtr)) return UIntPtrPrimitiveType;
+        else
```

Comment:
While I personally think that this looks somewhat reasonable, it is prohibited by our style guidelines: https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/coding-style.md, item 18, first bullet ("Never use single-line form")

Consider expression-switch

Reply:
@bartonjs I've tried using switch statement and I was not able to get it to compile (the compiler expect a constant value and `typeof(bool)` is not a const for it as of now).

Using braces turns this 22 lines into 77 lines and makes in unreadable.

According to https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/coding-style.md:

> Braces may be omitted only if the body of every block associated with an if/else if/.../else compound statement is placed on a single line.

So I've changed the last `else` block to be also a single line:

```diff
-else
-{
-    Debug.Fail($"{typeof(T).Name} was not expected.");
-    throw new InvalidOperationException();
-}
+else throw new InvalidOperationException();
```